osfoffline/filesystem_manager/sync_local_filesystem_and_db.py

```diff
@@ -46,42 +43,21 @@ def _make_local_db_tuple_list(self, local, db):
         if local and not isinstance(local, ProperPath):
             raise InvalidItemType
 
-        out = []
-        children = self._get_children(local) + self._get_children(db)
-        # sort in order to get matching items next to each other.
-        children = sorted(children, key=lambda c: self._get_proper_path(c).full_path)
-
-        i = 0
-        while i < len(children):
-            if self._represent_same_values(children, i):
-                if isinstance(children[i], Base):
-                    to_add = (children[i + 1], children[i])
-                else:
-                    to_add = (children[i], children[i + 1])
-                # add an extra 1 because skipping next value
-                i += 1
-            elif isinstance(children[i], Base):
-                to_add = (None, children[i])
-            else:
-                to_add = (children[i], None)
-            out.append(to_add)
-            i += 1
-
-        # assertions
-        for local, db in out:
-            if local is not None and db is not None:
-                assert isinstance(local, ProperPath)
-                assert isinstance(db, Base)
-                assert local == self._get_proper_path(db)
-            elif local is not None:
-                assert db is None
-                assert isinstance(local, ProperPath)
-            elif db is not None:
-                assert isinstance(db, Base)
-                assert local is None
-            else:
-                assert False
-        return out
+        local_children = {}
+        db_children = {}
+
+        for child in self._get_children(local):
+            assert isinstance(child, ProperPath)
+            local_children[child.full_path] = child
+
+        for child in self._get_children(db):
+            assert isinstance(child, Base)
+            db_children[self._get_proper_path(child).full_path] = child
+
+        return [
+            (local_children.get(path), db_children.get(path))
+            for path in set(list(local_children.keys()) + list(db_children.keys()))
```

Comment:
Not a required change, but one useful trick- instead of this creating two temporary lists and concatenating into a third temporary list, you can use `set(itertools.chain(d1.keys(), d2.keys()))`.

https://docs.python.org/3.5/library/itertools.html#itertools.chain